website/pages/guides/hcl/variables.mdx

```diff
@@ -18,54 +18,124 @@ Local variables can be a compound of input variables and local variables.
 
 ## Defining Variables and locals
 
-Let's create a file `variables.pkr.hcl` with the following contents.
+In the legacy JSON packer templates, any variables we hadn't already defined in
+the "variables" stanza of our json template could simply be passed in via the
+command line or a var-file; if a variable was never defined it would generally
+be interpolated to an empty string.
+
+*In the HCL2 packer templates, we must always pre-define our variables in the
+HCL equivalent of the "variables" stanza.*
+
+Another difference between JSON and HCL packer templates is that in JSON packer
+templates, the "variables" stanza, if used, was always in the same .json file
+as the builds and builders. In HCL, it can exist in its own file if you want it
+to.
+
+To demonstrate, let's create a file `variables.pkr.hcl` with the following
+contents.
 
 -> **Note**: that the file can be named anything, since Packer loads all
 files ending in `.pkr.hcl` in a directory. If you split your configuration
-across multiple files, use `packer build <source directory>` to initiate
-a build.
+across multiple files, use
+`packer build <command line flags> <source directory>` to initiate a build.
 
 ```hcl
-# variables.pkr.hcl
+//  variables.pkr.hcl
+
+// For those variables that you don't provide a default for, you must
+// set them from the command line, a var-file, or the environment.
+
+variable "weekday" {}
+
+variable "sudo_password" {
+  type =  string
+  default = "mypassword"
+  // Sensitive vars are hidden from output as of Packer v1.6.5
+  sensitive = true
+}
+
+variable "flavor" {
+  type = string
+  default = "strawberry"
+}
 
-variable "access_key" {}
-variable "secret_key" {}
-variable "region" {
-  default = "us-east-1"
+variable "exit_codes" {
+  type = list(number)
+  default = [0]
 }
 
 locals {
-  debian_ami_name = "${var.image_id}-debian"
+  ice_cream_flavor = "${var.flavor}-ice-cream"
   foo             = "bar"
 }
 ```
 
-This defines three variables within your Packer configuration. The first
-two have empty blocks `{}`. The third sets a default. If a default value is
-set, the variable is optional. Otherwise, the variable is required.
-This also defines two locals: `debian_ami_name` and `foo`.
+This defines several variables within your Packer configuration, each showing
+off a different way to set them. The first variable, "weekday", is an empty
+block `{}`. This will work under many simple circumstances, and Packer will
+guess what type the variable should be at runtime.
```

Comment:
This is not true. When the variable is passed from cli or env and we don't know it's type then we will expect it to be a `string`. But if it's passed from a varfile where we can interpret HCL properly it can be a slice or any supported type.